src/librustc_typeck/check/mod.rs

```diff
@@ -3211,6 +3218,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 *entry.get_mut() = adj;
             }
         }
+
+        // When there is an auto mutable borrow, it is equivalent to `&mut expr`,
+        // thus `expr` is ought to be typechecked with needs = [`Needs::MutPlace`].
+        // However in many cases it might not be checked this way originally, e.g.
+        // the receiver of a method call. We need to fix them up.
```

Comment:
This comments feels kinda sorta out of date? I guess `Needs` is still a thing, so never mind. But still, maybe we can say:

If there is an auto-mutable borrow, this is equivalent to adding `&mut <expr>`. This in turn may mean that some of the `Deref` or `Index` impls that appear in `<expr>` need to be rewritten to `DerefMut` or `IndexMut`. So go and fix those up.